Wrapper type for data that should never be logged

kylematsuda opened the pull request:
Fixes #11519.

So far this is just creating the new wrapper type. If this looks okay, I'll start adding this wrapper in places where tokens and secret keys are held or passed.

kylematsuda commented:
r? @Eh2406

Eh2406 commented:
This is more general than I had in mind. I would love to see what you end up doing with it. I was imagining a new type around `String`, so if your generic version gets difficult consider simplifying.

kylematsuda commented:
Thanks for taking a look! Sounds good, I'll give it a shot like this for now, but keeping the newtype approach in mind if things get hard (or if we decide that having this be generic is unnecessary).

kylematsuda commented:
I added the wrapper to most of the interfaces in `ops::registry` and `util::auth`. If this looks roughly like what you were envisioning, I'm happy to add this to other modules too as needed :smiley:

Eh2406 commented on src/bin/cargo/commands/login.rs:
```diff
@@ -36,7 +37,7 @@ pub fn cli() -> Command {
     ops::registry_login(
         config,
-        args.get_one("token").map(String::as_str),
+        args.get_one("token").map(String::as_str).map(Secret::from),
```
Maybe we should have a from &str, so that we can remove the repetition hear. What do you think?
(same for the other `.map(Secret::from)`)

kylematsuda replied:
Hmm, yes good point. Since `Secret<T>` impls `From<T>`, we can do `args.get_one::<String>("token").map(|s| s.as_str().into())`, which might be a little cleaner (that's what I did in the new commit). Or I can add a `fn from_str(&String) -> Secret<&str>` method if you think that'd be nicer :smile:

Eh2406 replied:
I was thinking of:

```
impl<T, U: From<T>> From<T> for Secret<U> {
    fn from(inner: T) -> Self {
        Self { inner.into() }
    }
}
```
but apparently that does not work.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b026fc5b25592691eb8fd44f4dd98452

kylematsuda replied:
Oh sorry, I see now. That would be nice, too bad it doesn't work :disappointed:

kylematsuda replied:
One option could be to remove the blanket `From<T>` impl and instead implement it individually for only the inner types that we actually wrap with `Secret`, which I think are currently only `String`, `&str`, and `AsymmetricKeyPair`. Then we can do `impl From<&String> for Secret<&str>`, etc.

(Sorry, I think I maybe got a little too excited and made things overly generic...)

Eh2406 commented on src/cargo/ops/registry.rs:
```diff
@@ -71,9 +71,9 @@ impl RegistryCredentialConfig {
         matches!(self, Self::AsymmetricKey(..))
     }
-    pub fn as_token(&self) -> Option<&str> {
+    pub fn as_token(&self) -> Option<Secret<&str>> {
```
Should this be `Secret<&str>` or `&Secret<String>`?
(same for the other `Secret<&str>`)

kylematsuda replied:
Hmm, I think I like `Secret<&str>` better, although of course I'm open to changing it. I went with `Option<Secret<&str>>` here since it was `Option<&str>` before. More generally, I guess I was thinking of `Secret<T>` as a wrapper similar to `Option<T>`, where I feel like I've seen `Option<&str>` a lot more than `&Option<String>` in fn signatures and such. Though to be honest I'm not sure I really understand the difference between those two types at a deep level. :sweat_smile: 

If you think that `Option<&Secret<String>>` is better, we might also want to consider changing the functions that currently take `Option<Secret<&str>>` as an argument to instead take `Option<&Secret<String>>`.

Eh2406 replied:
Sorry for not being clear. The `Option` is definitely required here. The question is whether we spell the borrowed form of `Secret<String>` as `&Secret<String>` or if we spell it `Secret<&str>`. As you point out, they're not significantly different types. Whatever we decide should be done consistently.

Eh2406 replied:
To both issues, it is hard to know the right level of "generic" until you've tried writing the code. Which you have done and I have not. I would appreciate if you would try `&Secret<String>`, I suspect that it will reduce some noise around these conversions and allow us to remove `Secret::as_deref`, but I don't know what other problems it may cause. I will then trust your judgment. After trying each, which one seems better to you?

Eh2406 replied:
Maybe we could make `map` more useful like:

```rust
    pub fn map<U, F>(&self, f: F) -> Secret<U>
    where
        F: FnOnce(&T) -> U,
    {
        Secret::from(f(&self.inner))
    }
```

Although maybe map isn't the right name for that new signature.

But either way is clearly fine. Whichever you prefer.

kylematsuda replied:
Let's go with `Secret<&str>` for now. Although it adds a bit of noise around conversions, I feel that it's a little simpler just by virtue of using similar patterns to other wrapper types like `Option` and `Result`.

Eh2406 commented on src/cargo/ops/registry.rs:
```diff
@@ -795,7 +795,8 @@ pub fn registry_login(
     let reg_cfg = auth::registry_credential_config(config, &source_ids.original)?;
 
-    let login_url = match registry(config, token, None, reg, false, None) {
+    let token = token.map(Secret::from);
```
Is this needed?

kylematsuda replied:
Good catch, thanks! I removed it

Eh2406 reviewed (approved):
Looks good! Ping me when you are ready for me to look again.

kylematsuda commented:
Hi! I changed back to using `Secret<&str>` and tried to improve the docs a bit. I also added a doctest to ensure that the `Debug` impl is working as expected.

I also noticed a few places where `token`s were being read or generated and then put inside a `Secret`, so I refactored those a little bit in the latest commit. I think it might be a little better, but also a little less readable, so I'd be interested to see what you think.

bors commented:
:pushpin: Commit efb972ae76c4e755b94466304c57dfc2b7a829d7 has been approved by `Eh2406`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=Eh2406 efb972ae76c4e755b94466304c57dfc2b7a829d7 -->
<!-- homu: {"type":"Approved","sha":"efb972ae76c4e755b94466304c57dfc2b7a829d7","approver":"Eh2406","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit efb972ae76c4e755b94466304c57dfc2b7a829d7 with merge d6ba7a3c5841d36b9d95d7e6a7481eba8109b018...
<!-- homu: {"type":"BuildStarted","head_sha":"efb972ae76c4e755b94466304c57dfc2b7a829d7","merge_sha":"d6ba7a3c5841d36b9d95d7e6a7481eba8109b018"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3964222343/jobs/6793270571)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/actions/runs/3964222343/jobs/6793270571","builder_name":"checks-actions"} -->

Eh2406 commented:
@bors retry

bors commented:
:hourglass: Testing commit efb972ae76c4e755b94466304c57dfc2b7a829d7 with merge ba6217e40fc05ac95f8a0fbe77aa83cdbc907a10...
<!-- homu: {"type":"BuildStarted","head_sha":"efb972ae76c4e755b94466304c57dfc2b7a829d7","merge_sha":"ba6217e40fc05ac95f8a0fbe77aa83cdbc907a10"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3964454095/jobs/6793697432)
Approved by: Eh2406
Pushing ba6217e40fc05ac95f8a0fbe77aa83cdbc907a10 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Eh2406","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/3964454095/jobs/6793697432"},"merge_sha":"ba6217e40fc05ac95f8a0fbe77aa83cdbc907a10"} -->
